corehq/apps/domain/views/settings.py

```diff
@@ -271,6 +273,103 @@ def post(self, request, *args, **kwargs):
         return self.get(request, *args, **kwargs)
 
 
+@method_decorator([
+    domain_admin_required,
+    use_bootstrap5,
+    toggles.IP_ACCESS_CONTROLS.required_decorator(),
+], name='dispatch')
+class EditIPAccessConfigView(BaseProjectSettingsView):
+    template_name = 'domain/admin/ip_access_config.html'
+    urlname = 'ip_access_config'
+    page_title = gettext_lazy("IP Access")
+
+    @property
+    @memoized
+    def form(self):
+        initial = {}
+        domain_config = self.ip_access_config
+        if domain_config:
+            display_spacer = ", "
+            initial.update({
+                'country_allowlist': domain_config.country_allowlist,
+                'ip_allowlist': display_spacer.join(domain_config.ip_allowlist),
+                'ip_denylist': display_spacer.join(domain_config.ip_denylist),
+                'comment': domain_config.comment
+            })
+
+        if self.request.method == 'POST':
+            return IPAccessConfigForm(self.request.POST, initial=initial)
+        return IPAccessConfigForm(initial=initial)
+
+    @cached_property
+    def ip_access_config(self):
+        try:
+            return IPAccessConfig.objects.get(domain=self.domain)
+        except IPAccessConfig.DoesNotExist:
+            return None
+
+    @cached_property
+    def ip_country(self):
+        if settings.MAXMIND_LICENSE_KEY:
+            return get_ip_country(self.current_ip)
+        return None
+
+    @property
+    def current_ip(self):
+        return get_ip(self.request)
+
+    @property
+    def page_context(self):
+        return {
+            'ip_access_config_form': self.form,
+        }
+
+    def post(self, request, *args, **kwargs):
+        if self.form.is_valid():
+            domain_config = self.ip_access_config
+            should_save = True
+            if not domain_config:
+                should_save = False
+                domain_config = IPAccessConfig()
+                domain_config.domain = self.domain
+            if not self.form.cleaned_data["ip_allowlist"] and not self.form.cleaned_data["country_allowlist"]:
+                should_save = False
+                messages.error(request,
+                               _("You must provide either your country or your "
+                                 "current IP Address to have access to this project."))
+            else:
+                for attr, value in self.form.cleaned_data.items():
+                    current_value = getattr(domain_config, attr)
+                    if value != current_value:
```

Comment:
This now means it only checks the country allowlist if it changes, and since validation happens inside here, it won't always be called.  You could set up a config which blocks your country but allows your IP, then on a subsequent request, delete your IP from the allowlist, causing you to lose access.

Reply:
oh good catch